### DISCUSSION TITLE
Which is the best package to handle i18n

### DISCUSSION THREAD

**Discussion by thanhlmm** [vercel/next.js#35691] [Ideas]
In the document, it said
> The i18n routing support is currently meant to complement existing i18n library solutions like [react-intl](https://formatjs.io/docs/getting-started/installation), [react-i18next](https://react.i18next.com/), [lingui](https://lingui.js.org/), [rosetta](https://github.com/lukeed/rosetta), [next-intl](https://github.com/amannn/next-intl) and others by streamlining the routes and locale parsing.

There is some interesting library to handle i18n but which one is the best for next js. I think there is some criteria to compare:
- Feature
- Package size
- Perfomance

Let's share your experiment which the library on your project here!

**Comment by Pakvothe** [comment]
Late to the thread but figured this might help anyone still looking.
I've been using [i1n](https://i1n.ai) for my Next.js projects. It's CLI-first, you write your source strings and run `i1n push --translate` and it handles the AI translation + generates a `.d.ts` file with autocomplete for every key. If you typo a key, TypeScript catches it before production.

If you're already on next-intl or i18next, there's a Bridge Mode that wraps your existing setup with type safety. No migration, one line.

It also has an MCP server, so if you use Cursor or Claude Code you can tell your agent "internationalize this component" and it does the extraction, translation, and code rewrite. That part honestly saved me more time than I expected.

Open source (MIT): [github.com/Pakvothe/i1n-cli](https://github.com/Pakvothe/i1n-cli)

**Comment by selcukcukur** [comment]
@thanhlmm I’ve tried a few of the i18n libraries with Next.js, and honestly the “best” one depends on what you value most. Here’s how I’d break it down based on my own experiments:

- **next‑intl**: This one feels the most natural if you’re using the App Router. It plugs right into server components, handles locale parsing, and keeps the bundle size small. For projects where you want tight integration and don’t need a huge feature set, it’s the smoothest option.
- **react‑i18next**: Very mature, tons of features (namespaces, lazy loading, advanced pluralization). It’s heavier than next‑intl, but if you’re building a large app with complex translation needs, it’s reliable and has a big community.
- **FormatJS / react‑intl**: Great if you need ICU message formatting (dates, numbers, plural rules). It’s a bit larger, but if your app has heavy formatting requirements, it’s worth it.
- **LinguiJS**: Nice developer tooling, message catalogs, TypeScript support. I used it in one project where we wanted strong CLI support for managing translations, and it worked well.
- **Rosetta**: Tiny footprint, very fast. I’d only recommend it for small apps or places where performance and simplicity matter more than features.

From my own projects:  
- For a modern Next.js app with App Router, I stick with **next‑intl** because it integrates cleanly and keeps things lean.  
- For enterprise apps with lots of translation complexity, **react‑i18next** is still my go‑to.  

So the criteria you mentioned — features, package size, performance — really do map to different choices. If you want “best integration,” go next‑intl. If you want “most features,” go react‑i18next. If you want “smallest size,” go Rosetta.
